File: 30 Days of Code/Python/19 - Day 18 - Queues and Stacks.py
# ========================
#       Information
# ========================

# Direct Link: https://www.hackerrank.com/challenges/30-queues-stacks/problem
# Difficulty: Easy
# Max Score: 30
# Language: Python

# ========================
#         Solution
# ========================

import sys # Solution made without using the required library

class Solution:
    # Write your code here
    def __init__(self):
        self.top = 0
        self.rear = 0
        self.a = list()
        self.r = list()

    def pushCharacter(self, char):
        self.a.append(char)
        self.top += 1

    def enqueueCharacter(self, char):
        self.r.append(char)
        self.rear += 1

    def popCharacter(self):
        self.top -= 1
        return self.a[self.top]

    def dequeueCharacter(self):
        return self.r[i]

# Read the string s
s = input()

# Create the Solution class object
obj = Solution()

l = len(s)

# Push/enqueue all the characters of string s to stack
for i in range(l):
    obj.pushCharacter(s[i])
    obj.enqueueCharacter(s[i])

isPalindrome = True

'''
pop the top character from stack
dequeue the first character from queue
compare both the characters
'''

for i in range(l // 2):
    if obj.popCharacter() != obj.dequeueCharacter():
        isPalindrome = False
        break

#finally print whether string s is palindrome or not.
if isPalindrome:
    print("The word, "+s+", is a palindrome.")
else:
    print("The word, "+s+", is not a palindrome.")

File: 30 Days of Code/C#/19 - Day 18 - Queues and Stacks.cs
// ========================
//       Information
// ========================

// Direct Link: https://www.hackerrank.com/challenges/30-queues-stacks/problem
// Difficulty: Easy
// Max Score: 30
// Language: C#

// ========================
//         Solution
// ========================

using System;
using System.Collections.Generic;
using System.IO;

class Solution {
	public Stack < char > charS = new Stack < char > ();
	public Queue < char > charQ = new Queue < char > ();

	void pushCharacter(char c) {
		charS.Push(c);
	}

	char popCharacter() {
		return charS.Pop();
	}
	void enqueueCharacter(char c) {
		charQ.Enqueue(c);
	}

	char dequeueCharacter() {
		return charQ.Dequeue();
	}

	static void Main(String[] args) {
		// read the string s.
		string s = Console.ReadLine();

		// create the Solution class object p.
		Solution obj = new Solution();

		// push/enqueue all the characters of string s to stack.
		foreach(char c in s) {
			obj.pushCharacter(c);
			obj.enqueueCharacter(c);
		}

		bool isPalindrome = true;

		// pop the top character from stack.
		// dequeue the first character from queue.
		// compare both the characters.
		for (int i = 0; i < s.Length / 2; i++) {
			if (obj.popCharacter() != obj.dequeueCharacter()) {
				isPalindrome = false;

				break;
			}
		}
		// finally print whether string s is palindrome or not.
		if (isPalindrome) {
			Console.Write("The word, {0}, is a palindrome.", s);
		} else {
			Console.Write("The word, {0}, is not a palindrome.", s);
		}
	}
}
